### PR TITLE
Support TFJob(kubeflow) in Multikueue

### LINKED ISSUE #2552: Support Kubeflow Jobs in MultiKueue
**What would you like to be added**:

Support for Kubeflow Jobs in MultiKueue, in particular, for TFJob and PyTorchJob.
Ideally, the implementation should be mostly common among all job types.

Kubeflow Job doesn't have support for managedBy, so, for now, we can only support the scenario where the manager cluster doesn't have the controller installed.

**Why is this needed**:

To continue the incremental improvement of MK and satisfy the needs of early adopters.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by mszadkow**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
The PR introduces a new Multikueue adapter to handle TFJob (Kubeflow).
We want to extend Multikueue capabilities to satisfy the needs of early adopters.

#### Which issue(s) this PR fixes:

Relates #2552

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/admissionchecks/multikueue/indexer_test.go [hunk 1]
```diff
@@ -47,6 +48,7 @@ func getClientBuilder() (*fake.ClientBuilder, context.Context) {
 	utilruntime.Must(kueue.AddToScheme(scheme))
 	utilruntime.Must(kueuealpha.AddToScheme(scheme))
 	utilruntime.Must(jobset.AddToScheme(scheme))
+	utilruntime.Must(kftraining.AddToScheme(scheme))
```
[the remark this answers is not in the document]
Comment (mszadkow) [reply pkg/controller/admissionchecks/multikueue/indexer_test.go:51]: test fail if we don't provide the scheme with:
```golangmultikueuecluster_test.go:421: unexpected controllers (-want/+got):

**Review by trasc — COMMENTED** [review]
/lgtm

File: pkg/controller/jobframework/validation.go [hunk 2]
```diff
@@ -35,8 +36,10 @@ var (
 	labelsPath                    = field.NewPath("metadata", "labels")
 	queueNameLabelPath            = labelsPath.Key(constants.QueueLabel)
 	workloadPriorityClassNamePath = labelsPath.Key(constants.WorkloadPriorityClassLabel)
-	supportedPrebuiltWlJobGVKs    = sets.New(batchv1.SchemeGroupVersion.WithKind("Job").String(),
-		jobset.SchemeGroupVersion.WithKind("JobSet").String())
+	supportedPrebuiltWlJobGVKs    = sets.New(
+		batchv1.SchemeGroupVersion.WithKind("Job").String(),
+		jobset.SchemeGroupVersion.WithKind("JobSet").String(),
+		kftraining.SchemeGroupVersion.WithKind("TFJob").String())
```
Comment (tenzen-y) [inline pkg/controller/jobframework/validation.go:42]:
```suggestion
		kftraining.SchemeGroupVersion.WithKind(kftraining.TFJobKind).String())
```

File: hack/e2e-common.sh [hunk 3]
```diff
@@ -23,6 +23,12 @@ export JOBSET_MANIFEST=https://github.com/kubernetes-sigs/jobset/releases/downlo
 export JOBSET_IMAGE=registry.k8s.io/jobset/jobset:${JOBSET_VERSION}
 export JOBSET_CRDS=${ROOT_DIR}/dep-crds/jobset-operator/
 
+export KUBEFLOW_VERSION=v1.7.0
+export KUBEFLOW_MANIFEST=https://github.com/kubeflow/training-operator/manifests/overlays/standalone?ref=${KUBEFLOW_VERSION}
+#no matching semver tag unfortunately
+export KUBEFLOW_IMAGE=kubeflow/training-operator:v1-855e096
```
Comment (tenzen-y) [inline hack/e2e-common.sh:29]: Could you open an issue so that we can modify this script once https://github.com/kubeflow/training-operator/issues/2155 is resolved?

File: pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	remoteJob := &kftraining.TFJob{}
+	err = remoteClient.Get(ctx, key, remoteJob)
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// if the remote exists, just copy the status
+	if err == nil {
+		localJob.Status = remoteJob.Status
+		return localClient.Status().Update(ctx, &localJob)
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go:57]:
```suggestion
		return clientutil.PatchStatus(ctx, localClient, &localJob, func() (bool, error) {
			localJob.Status = remoteJob.Status
			return true, nil
		})
```

Shouldn't the patch operation be used to avoid unintentional dropped fields?

File: pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	remoteJob = &kftraining.TFJob{
+		ObjectMeta: api.CloneObjectMetaForCreation(&localJob.ObjectMeta),
+		Spec:       *localJob.Spec.DeepCopy(),
+	}
+
+	// add the prebuilt workload
+	if remoteJob.Labels == nil {
+		remoteJob.Labels = map[string]string{}
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go:67]:
```suggestion
		remoteJob.Labels = make(map[string]string, 2)
```

File: pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	return client.IgnoreNotFound(remoteClient.Delete(ctx, &job))
+}
+
+func (b *multikueueAdapter) KeepAdmissionCheckPending() bool {
+	return false
+}
+
+func (b *multikueueAdapter) IsJobManagedByKueue(ctx context.Context, c client.Client, key types.NamespacedName) (bool, string, error) {
+	return true, "", nil
+}
```
Comment (tenzen-y) [inline pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go:90]: Uhm, if users accidentally install the training-operator to management cluster, I guess that operation conflicting between the kueue-manager and the training-operator, right?
But I understand there are no solutions to avoid the situation automatically.

So, mentioning this limitation in the documentation would be worth it.

File: test/e2e/multikueue/e2e_test.go [hunk 7]
```diff
@@ -364,6 +367,118 @@ var _ = ginkgo.Describe("MultiKueue", func() {
...
+		ginkgo.It("Should run a kubeflow TFJob on worker if admitted", func() {
+			tfJob := testingtfjob.MakeTFJob("tfjob1", managerNs.Name).
+				Queue(managerLq.Name).
+				TFReplicaSpecs(
+					testingtfjob.TFReplicaSpecRequirement{
+						ReplicaType:  kftraining.TFJobReplicaTypeChief,
+						ReplicaCount: 1,
+						Annotations: map[string]string{
+							"sidecar.istio.io/inject": "false",
+						},
```
Comment (tenzen-y) [inline test/e2e/multikueue/e2e_test.go:379]: This `istio` annotation is not needed since this annotation is used for the k8s cluster with istio.

File: test/e2e/multikueue/e2e_test.go [hunk 8]
```diff
@@ -364,6 +367,118 @@ var _ = ginkgo.Describe("MultiKueue", func() {
...
+						ReplicaType:  kftraining.TFJobReplicaTypePS,
+						ReplicaCount: 1,
+						Annotations: map[string]string{
+							"sidecar.istio.io/inject": "false",
+						},
+						RestartPolicy: "Never",
+					},
+					testingtfjob.TFReplicaSpecRequirement{
+						ReplicaType:  kftraining.TFJobReplicaTypeWorker,
+						ReplicaCount: 2,
```
Comment (tenzen-y) [inline test/e2e/multikueue/e2e_test.go:392]:
```suggestion
						ReplicaCount: 1,
```

Only 1 replica should be enough, right?

File: test/integration/multikueue/multikueue_test.go [hunk 9]
```diff
@@ -769,6 +772,160 @@ var _ = ginkgo.Describe("Multikueue", ginkgo.Ordered, ginkgo.ContinueOnFailure,
...
+			}, kueue.PodSetAssignment{
+				Name: "worker",
+			},
+		).Obj()
+
+		ginkgo.By("setting workload reservation in the management cluster", func() {
+			createdWorkload := &kueue.Workload{}
+			gomega.Eventually(func(g gomega.Gomega) {
+				g.Expect(managerTestCluster.client.Get(managerTestCluster.ctx, wlLookupKey, createdWorkload)).To(gomega.Succeed())
+				fmt.Println(createdWorkload.Spec.PodSets)
```
Comment (tenzen-y) [inline test/integration/multikueue/multikueue_test.go:817]:
```suggestion
```

Is this a debug code?

File: test/e2e/multikueue/e2e_test.go [hunk 10]
```diff
@@ -364,6 +367,118 @@ var _ = ginkgo.Describe("MultiKueue", func() {
...
+				).
+				Request(kftraining.TFJobReplicaTypePS, corev1.ResourceCPU, "0.5").
+				Request(kftraining.TFJobReplicaTypePS, corev1.ResourceMemory, "200M").
+				Request(kftraining.TFJobReplicaTypePS, corev1.ResourceCPU, "0.5").
+				Request(kftraining.TFJobReplicaTypePS, corev1.ResourceMemory, "200M").
+				Request(kftraining.TFJobReplicaTypeWorker, corev1.ResourceCPU, "0.5").
+				Request(kftraining.TFJobReplicaTypeWorker, corev1.ResourceMemory, "100M").
+				Image(kftraining.TFJobReplicaTypeChief, "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"5s"}).
+				Image(kftraining.TFJobReplicaTypePS, "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"5s"}).
+				Image(kftraining.TFJobReplicaTypeWorker, "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"5s"}).
```
Comment (mimowo) [inline test/e2e/multikueue/e2e_test.go:407]: No strong view here, but I think the 5s delays for the e2e tests will not scale too well if we do it per integration. IIUC this is to test the live-updates, but maybe we could rely on integration tests for the  feature, where we have more control on timings? IIUC live updates are covered by the integration test: "changing the status of the TFJob in the worker, updates the manager's TFJob status".  I'm also a bit concerned with potential for flakes here, so I would prefer 1ms images for the baseline scenario, and just check here that the TFJob is finished.

File: hack/e2e-common.sh [hunk 11]
```diff
@@ -23,6 +23,12 @@ export JOBSET_MANIFEST=https://github.com/kubernetes-sigs/jobset/releases/downlo
 export JOBSET_IMAGE=registry.k8s.io/jobset/jobset:${JOBSET_VERSION}
 export JOBSET_CRDS=${ROOT_DIR}/dep-crds/jobset-operator/
 
+export KUBEFLOW_VERSION=v1.7.0
```
[the remark this answers is not in the document]
Comment (mszadkow) [reply hack/e2e-common.sh:26]: sure we can

File: hack/e2e-common.sh (hunk 3 above)
Comment (mszadkow) [reply hack/e2e-common.sh:29]: Good idea

Comment (mszadkow) [reply hack/e2e-common.sh:29]: Done: https://github.com/kubernetes-sigs/kueue/issues/2709

File: pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go (hunk 4 above)
Comment (mszadkow) [reply pkg/controller/jobs/kubeflow/jobs/tfjob/tfjob_multikueue_adapter.go:57]: good point

File: test/e2e/multikueue/e2e_test.go (hunk 10 above)
Comment (mszadkow) [reply test/e2e/multikueue/e2e_test.go:407]: 3 more integrations to go, so I fully agree on scalability side.
Let's decrease the delays and check status in integration tests

**Comment by mimowo** [comment]
@mszadkow please ping us in a comment when the PR is ready for the second pass after the updates.

**Comment by mszadkow** [comment]
@tenzen-y @mimowo 
I think it's ready

File: pkg/controller/jobs/jobset/jobset_multikueue_adapter.go [hunk 12]
```diff
@@ -68,7 +68,7 @@ func (b *multikueueAdapter) SyncJob(ctx context.Context, localClient client.Clie
 
 	// add the prebuilt workload
 	if remoteJob.Labels == nil {
-		remoteJob.Labels = map[string]string{}
+		remoteJob.Labels = make(map[string]string, 2)
```
Comment (trasc) [inline pkg/controller/jobs/jobset/jobset_multikueue_adapter.go:71]: this is not related to tf ob, I'd prefer to do it in a clean-up PR.

File: site/content/en/docs/tasks/manage/setup_multikueue.md [hunk 13]
```diff
@@ -74,6 +74,16 @@ kubectl apply --server-side -f https://raw.githubusercontent.com/kubernetes-sigs
 ```
 {{% /alert %}}
 
+### Kubeflow Installation
```
Comment (trasc) [inline site/content/en/docs/tasks/manage/setup_multikueue.md:77]: Add a link to the kubeflow installation guide, to be done in the worker clusters.

File: test/e2e/multikueue/e2e_test.go [hunk 14]
```diff
@@ -364,6 +367,87 @@ var _ = ginkgo.Describe("MultiKueue", func() {
...
+					}, util.IgnoreConditionTimestampsAndObservedGeneration))
+					g.Expect(workload.FindAdmissionCheck(createdWorkload.Status.AdmissionChecks, multiKueueAc.Name)).To(gomega.BeComparableTo(&kueue.AdmissionCheckState{
+						Name:    multiKueueAc.Name,
+						State:   kueue.CheckStateReady,
+						Message: `The workload got reservation on "worker1"`,
+					}, cmpopts.IgnoreFields(kueue.AdmissionCheckState{}, "LastTransitionTime")))
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("Waiting for the TfJob to finish", func() {
```
Comment (trasc) [inline test/e2e/multikueue/e2e_test.go:426]: check the final status of the job also

**Review by trasc — COMMENTED** [review]
some nits, otherwise LGTM

File: site/content/en/docs/concepts/multikueue.md [hunk 15]
```diff
@@ -68,6 +68,12 @@ The `managedBy` field is available as an Alpha feature staring Kubernetes 1.30.0
 
 We recommend using JobSet v0.5.1 or newer.
 
+### Kubeflow
+
+The supported version of the Kubeflow Training Operator is v1.7.0.
```
Comment (tenzen-y) [inline site/content/en/docs/concepts/multikueue.md:73]:
```suggestion
The supported version of the Kubeflow Training Operator is v1.7.0, or a newer version. 
```

File: site/content/en/docs/tasks/manage/setup_multikueue.md [hunk 16]
```diff
@@ -49,6 +49,11 @@ chmod +x create-multikueue-kubeconfig.sh
 
 To create a Kubeconfig that can be used in the manager cluster to delegate Jobs in the current worker.
 
+### Kubeflow Installation
+
+Install Kubeflow Training-operator in the Worker cluster (see [Kubeflow Training-operator Installation](https://www.kubeflow.org/docs/components/training/installation/)
+for more details). Please use version v1.7.0 for MultiKueue.
```
Comment (tenzen-y) [inline site/content/en/docs/tasks/manage/setup_multikueue.md:55]:
```suggestion
for more details). Please use version v1.7.0 or a newer version for MultiKueue.
```

File: site/content/en/docs/tasks/manage/setup_multikueue.md [hunk 17]
```diff
@@ -74,6 +79,16 @@ kubectl apply --server-side -f https://raw.githubusercontent.com/kubernetes-sigs
 ```
 {{% /alert %}}
 
+### Kubeflow Installation
+
+{{% alert title="Warning" color="warning" %}}
+Make sure to only install the Kubeflow TFJobs CRD of version v1.7.0 on the management cluster.
```
Comment (tenzen-y) [inline site/content/en/docs/tasks/manage/setup_multikueue.md:85]:
```suggestion
Make sure to install only the Kubeflow TFJobs CRD of version v1.7.0 or newer on the management cluster.
```

File: site/content/en/docs/tasks/manage/setup_multikueue.md [hunk 18]
```diff
@@ -74,6 +79,16 @@ kubectl apply --server-side -f https://raw.githubusercontent.com/kubernetes-sigs
...
+
+```bash
+  kubectl apply --server-side -f https://github.com/kubeflow/training-operator/blob/v1.7.0/manifests/base/crds/kubeflow.org_tfjobs.yaml
```
Comment (tenzen-y) [inline site/content/en/docs/tasks/manage/setup_multikueue.md:88]:
```suggestion
  kubectl apply --server-side -f https://github.com/kubeflow/training-operator/blob/v1.8.0/manifests/base/crds/kubeflow.org_tfjobs.yaml
```

Let's use the latest Kubeflow Training Operator version.

**Review by tenzen-y — COMMENTED** [review]
Some minor change suggestions.
/approve

**Review by trasc — COMMENTED** [review]
/lgtm

**Comment by tenzen-y** [comment]
/release-note-edit

```release-note
MultiKueue: Support for the Kubeflow TFJob
```
